### PR TITLE
Cache latest version

### PR DESCRIPTION
**Pull request by amcaplan**
### WHY are these changes introduced?

We make a call to npm each CLI run when checking for versions. This is wasteful and time-consuming.

### WHAT is this pull request doing?

Makes it faster! We still pay full price on the initial run, but afterwards will save time and frustration by caching the result.

This is a generalizable system for any similarly expensive operation that doesn't need to be repeated more than once daily (or another amount of time of our choice).

### How to test your changes?

`dev shopify version`

It should still work, and get faster!

### Measuring impact

How do we know this change was effective? Please choose one:

- [x] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes
- [x] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

**Review by matteodepalo — APPROVED** [review]
Nice! 💯

File: packages/cli-kit/src/version.ts [hunk 1]
```diff
@@ -9,9 +10,10 @@ import latestVersion from 'latest-version'
  * @param name - The name of the NPM package.
  * @returns A promise to get the latest available version of a package.
  */
-export async function latestNpmPackageVersion(name: string) {
+export async function latestNpmPackageVersion(name: string, options = {force: false}) {
   debug(content`Getting the latest version of NPM package: ${token.raw(name)}`)
-  return latestVersion(name)
+  if (options.force) return latestVersion(name)
```
Comment (shauns) [inline packages/cli-kit/src/version.ts:15]: If I force load this, shouldn't the cache be updated?

File: packages/cli-kit/src/store.ts [hunk 2]
```diff
@@ -18,6 +18,7 @@ interface ConfSchema {
   appInfo: CachedAppInfo[]
   themeStore: string
   session: string
+  cache: {[key: string]: {value: unknown; updatedAt: string}}
```
Comment (shauns) [inline packages/cli-kit/src/store.ts:21]: Shouldn't `cache` be in the `schema` below?

File: packages/cli-kit/src/version.ts [hunk 3]
```diff
@@ -9,9 +10,10 @@ import latestVersion from 'latest-version'
...
+  return cache(`latest-npm-package-version-${name}`, async () => latestVersion(name))
```
Comment (shauns) [inline packages/cli-kit/src/version.ts:16]: I think it we're doing this it needs to be clearer that `cache` is a system-wide thing, not local to a project or in-memory on a process.

File: packages/cli-kit/src/store.ts [hunk 4]
```diff
@@ -183,4 +193,18 @@ export class CLIKitStore extends Conf<ConfSchema> {
     debug(content`Removing session store...`)
     this.set('sessionStore', '')
   }
+
+  async cache<TValue>(key: string, source: () => Promise<TValue>, expirySeconds = 24 * 3600): Promise<TValue> {
+    const expiryMilliseconds = expirySeconds * 1000
+    const cache = this.get('cache') || {}
+    if (cache[key]) {
```
Comment (shauns) [inline packages/cli-kit/src/store.ts:200]: if you pull `cache[key]` into a variable before hitting the `if`, TS's type narrowing should mean you need a lot fewer `!`.

**Comment by amcaplan** [comment]
I think this PR is hopelessly out of date with how CLI-kit is organized. Closing, hoping to reopen a better version.

**Pull request closed without merging**